src/box/iproto.cc

```diff
@@ -3041,36 +3090,92 @@ iproto_free(void)
 	evio_service_stop(&tx_binary);
 }
 
-int
-iproto_rmean_foreach(void *cb, void *cb_ctx)
+static void
+iproto_thread_rmean(struct iproto_thread *iproto_thread,
+		    int64_t *mean, int64_t *total, size_t name)
+{
+	assert(mean != NULL && total != NULL);
+	*mean += rmean_mean(iproto_thread->rmean, name);
+	*total += rmean_total(iproto_thread->rmean, name);
+}
+
+static void
+iproto_thread_tx_rmean(struct iproto_thread *iproto_thread,
+		       int64_t *mean, int64_t *total, size_t name)
+{
+	assert(mean != NULL && total != NULL);
+	*mean += rmean_mean(iproto_thread->tx.rmean, name);
+	*total += rmean_total(iproto_thread->tx.rmean, name);
+}
+
+typedef void (*rmean_stats_cb)(struct iproto_thread *iproto_thread,
+			       int64_t *mean, int64_t *total, size_t name);
+
+static int
+iproto_thread_rmean_foreach(const char **strings, rmean_stats_cb stats_cb,
+			    int thread_id, size_t rmean_last_name,
+			    void *cb, void *cb_ctx)
 {
-	for (size_t i = 0; i < RMEAN_NET_LAST; i++) {
+	int rc = 0;
+	for (size_t i = 0; i < rmean_last_name; i++) {
 		int64_t mean = 0;
 		int64_t total = 0;
-		for (int j = 0; j < iproto_threads_count; j++)  {
-			mean += rmean_mean(iproto_threads[j].rmean, i);
-			total += rmean_total(iproto_threads[j].rmean, i);
-		}
-		int rc = ((rmean_cb)cb)(rmean_net_strings[i], mean,
-					total, cb_ctx);
+		stats_cb(&iproto_threads[thread_id], &mean, &total, i);
+		if (((rmean_cb)cb)(strings[i], mean, total, cb_ctx) != 0)
+			rc = 1;
+	}
+	return rc;
+}
+
+static int
+iproto_rmean_foreach(const char **strings, rmean_stats_cb stats_cb,
+		     size_t rmean_last_name, void *cb, void *cb_ctx)
+{
+	for (size_t i = 0; i < rmean_last_name; i++) {
+		int64_t mean = 0;
+		int64_t total = 0;
+		for (int j = 0; j < iproto_threads_count; j++)
+			stats_cb(&iproto_threads[j], &mean, &total, i);
+		int rc = ((rmean_cb)cb)(strings[i], mean, total, cb_ctx);
 		if (rc != 0)
 			return rc;
 	}
 	return 0;
 }
 
+int
+iproto_rmean_foreach(void *cb, void *cb_ctx)
+{
+	int rc;
+	if ((rc = iproto_rmean_foreach(rmean_net_strings,
+				       iproto_thread_rmean,
+				       RMEAN_NET_LAST,
+				       cb, cb_ctx)) != 0)
+		return rc;
+	if ((rc = iproto_rmean_foreach(rmean_tx_strings,
+				       iproto_thread_tx_rmean,
+				       RMEAN_TX_LAST,
+				       cb, cb_ctx)) != 0)
```

Comment:
You can pass an rmean object instead of the first three fields. An rmean object contains the names and the number of stats.

Reply:
For `iproto_thread_rmean_foreach` you right, but in this fucntion there is no rmean object, we should iterate over all threads and collect stats for there rmeans.

Reply:
We can use offsetof to calculate what rmean we want to pass.

Reply:
Maybe, pass the offset of rmean in iproto_thread then (see offsetof)?

Reply:
Missed you comment and suggested the same. Let's try that.

Reply:
fixed